Improve auth validation

carlobeltrame opened the pull request:
Improves frontend validations in the auth views.
"Fixes" password confirmation validation on password change view, but you can still submit the form with invalid data...
Works toward #2563.

carlobeltrame committed:
Validate email on login form

carlobeltrame committed:
Unify usage of required validation

carlobeltrame committed:
Translate and strengthen password validation

Refs #2563

usu commented on frontend/src/views/auth/ResetPassword.vue:
```diff
@@ -56,9 +56,8 @@
           :label="$tc('views.auth.resetPassword.passwordConfirmation')"
           name="Confirmation"
-          vee-id="confirmation"
-          vee-rules="pwConfirmed:@password"
```
What exactly was not working for you?

On dev deployment it's validating the field properly with the `pwConfirmed` rule:

<img width="592" alt="image" src="https://user-images.githubusercontent.com/449555/194584179-b688ca61-5313-4e62-ac28-9f52e5b1ea63.png">

carlobeltrame replied:
It was not working on Register.vue (because of a missing validation-observer, which I didn't realize at the time), and I wanted to have it consistent between that page and ResetPassword.vue. I now removed our custom rule `pwConfirmed` completely and returned to the vee-validate built-in `confirmed`.

usu commented on frontend/src/views/auth/Register.vue:
```diff
@@ -5,49 +5,48 @@
+        vee-rules="required"
         append-icon="mdi-account-outline"
         dense
-        required
```
This breaks the red-star-styling that was introduced by @pmattmann in https://github.com/ecamp/ecamp3/pull/2845

carlobeltrame replied:
That is the reason why I am setting the required attribute on ValidationProvider. ValidationProvider does not do anything with it, but the CSS relies on a parent element having that attribute.

carlobeltrame replied:
More generally, I'd like to propose that we don't use the `required` attribute directly, but rather keep all validation rules in one single property `vee-rules`. If we use `required`, we will at some point start using `pattern` and other attributes as well, and it's going to be a mess. `required` is an attribute on the native HTML `<input>` element, we don't have to pretend our `<e-text-field>` is an input, because it is not.

usu commented on frontend/src/components/form/base/ETextField.vue:
```diff
@@ -6,11 +6,13 @@
     :vid="veeId"
     :rules="veeRules"
     class="e-form-container"
+    :required="required"
```
I think there's no property named `required` on `ValidationProvider`:
https://vee-validate.logaretm.com/v3/api/validation-provider.html#props

What is this expected to do?

carlobeltrame replied:
See above, necessary for the asterisk. In the meantime, I added it to all e-inputs.

usu commented on frontend/src/views/auth/Register.vue:
```diff
@@ -5,49 +5,48 @@
       <e-text-field
         v-model="firstname"
         :name="$tc('entity.user.fields.firstname')"
+        vee-rules="required"
```
Theoretically, vee-validate should be able to infer the required rule from the required html attribute. If this doesn't work, might be worth to dig a bit on why this is the case.

https://vee-validate.logaretm.com/v3/guide/rules.html#required

carlobeltrame replied:
See above, I'd propose to keep everything in `vee-rules`

usu commented on frontend/src/views/auth/ResetPassword.vue:
```diff
@@ -56,9 +56,8 @@
-          vee-rules="pwConfirmed:@password"
-          vee-rules-old="confirmed:password"
+          vee-rules="required"
+          :rules="pw2Rules"
```
I'd tend to avoid mixing vuetify validation and vee-validate, unless absolutely necessary.

carlobeltrame replied:
Agreed, in the meantime I was able to return to native vee-validate only.

carlobeltrame committed:
Pass required validation flag to all e-inputs

carlobeltrame committed:
Fix vee-validate built-in rule by introducing a validation-observer

carlobeltrame committed:
Use built-in vee-validate rules whenever possible

carlobeltrame committed:
Adjust register and password reset forms to look more similar

carlobeltrame committed:
Remove obsolete translations

carlobeltrame committed:
Merge remote-tracking branch 'origin/devel' into login-email-validation

BacLuc commented on api/tests/Api/ResetPassword/UpdatePasswordTest.php:
```diff
@@ -0,0 +1,112 @@
+    // TODO tests for recaptcha check
+    // TODO tests for requesting a password reset and retrieving the password reset data, in separate test files
+
+    protected ?User $user = null;
+    protected ?string $passwordResetKey = null;
+    protected $client;
```
private would be enough

BacLuc commented on api/tests/Api/ResetPassword/UpdatePasswordTest.php:
```diff
@@ -0,0 +1,112 @@
+    // TODO tests for recaptcha check
+    // TODO tests for requesting a password reset and retrieving the password reset data, in separate test files
+
+    protected ?User $user = null;
```
must this field be nullable?
should we do this in all tests and set the field initially to null?

carlobeltrame replied:
I think it has to be nullable because I don't fill it in the constructor. But I might be wrong. I assigned null because then the error wouldn't be as ugly if I tried to access $this->user before assigning to it. No particularly good reason, I know.

BacLuc commented on api/tests/Api/ResetPassword/UpdatePasswordTest.php:
```diff
@@ -0,0 +1,112 @@
+    public function testPatchResetPasswordAllowsLongPassword() {
+        $this->mockRecaptcha();
+        $this->client->request('PATCH', '/auth/reset_password/'.$this->passwordResetKey, ['json' => [
+            'password' => 'this password has a total of 122 characters. this password has a total of 122 characters. OWASP approves of this password.',
```
i used str_repeat to generate these long strings.
makes it easier if we change the length

BacLuc commented on api/tests/Api/ResetPassword/UpdatePasswordTest.php:
```diff
@@ -0,0 +1,112 @@
+    public function testPatchResetPasswordValidatesUnreasonablyLongPassword() {
+        $this->mockRecaptcha();
+        $this->client->request('PATCH', '/auth/reset_password/'.$this->passwordResetKey, ['json' => [
+            'password' => 'this password has a total of more than 128 characters. this password has a total of more than 128 characters. OWASP does not approve this password.',
```
i used str_repeat to generate these long strings.
makes it easier if we change the length

BacLuc commented on api/tests/Api/ResetPassword/UpdatePasswordTest.php:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    public function testPatchResetPasswordValidatesUnreasonablyLongPassword() {
+        $this->mockRecaptcha();
```
why not call this function directly in the setup?

carlobeltrame replied:
Because I anticipate there will be a test which will call it like `$this->mockRecaptcha(false)` in order to simulate a failed reCaptcha check and test what happens then.

BacLuc commented on frontend/src/mixins/passwordStrengthMixin.js:
```diff
@@ -0,0 +1,21 @@
+import { passwordStrength } from 'check-password-strength'
+
+export const passwordStrengthMixin = {
```
maybe Composition Api, or one self contained component

BacLuc reviewed (approved):
only minor comments

carlobeltrame commented on frontend/src/components/form/base/EDatePicker.vue:
```diff
@@ -10,6 +10,7 @@ Displays a field as a date picker (can be used with v-model)
     :format-picker="formatPicker"
     :parse="parse"
     :parse-picker="parsePicker"
+    :required="required"
```
FIXME: This component and some others don't use the formComponentPropsMixin

carlobeltrame committed:
Move  computation to separate mixin which can be used in all e-input components

carlobeltrame committed:
Run linter

carlobeltrame committed:
Fix tests

carlobeltrame committed:
Slight refactoring from review

usu commented on frontend/src/views/auth/Register.vue:
```diff
@@ -2,98 +2,109 @@
-        <v-spacer />
-        <icon-spacer />
-      </v-btn>
+        <v-btn type="submit" color="primary" :disabled="!formComplete" block x-large>
```
Brauchen wir `:disabled="!formComplete"` noch, jetzt wo wir vernünftige validations haben und einen `ValidationObserver`?

Das sonst übliche Verhalten ist ja eigentlich, dass der Submit-button aktiviert ist, wenn ich darauf klicke werden aber alle Feldvalidierungen angestossen.

Es ist im Zustand as-is relativ einfach, die Checkbox mit den Nutzungsbedingungen zu übersehen und dann nicht auf Anhieb zu verstehen, wieso der Register-Button nicht funktioniert.

carlobeltrame replied:
I tried hard to use validation-observer for everything, but I could not get it to validate the checkbox on submit. Everything else is validated through the validation-observer now in 867e7434fb9e979f998422eadb04ad950c592ed7, but for that checkbox I still had to keep formComplete. Would need very deep debugging of vee-validate and/or vuetify to fix that.

usu commented on frontend/src/views/auth/ResetPassword.vue:
```diff
@@ -38,28 +38,33 @@
         <e-text-field
           v-model="password"
           :label="$tc('views.auth.resetPassword.password')"
-          name="Password"
```
name should be populated and label optionally, not the other way round:
https://github.com/ecamp/ecamp3/issues/2367

Otherwise vee-validate doesn't know the field name.

![image](https://user-images.githubusercontent.com/449555/194693525-72271920-de1d-4745-a49e-ce13f8b02d10.png)

carlobeltrame replied:
Done in efb4210998301a890f3056632f7010e82f9532a6

usu commented on frontend/src/views/auth/ResetPassword.vue:
```diff
@@ -38,28 +38,33 @@
         <e-text-field
           v-model="confirmation"
           :label="$tc('views.auth.resetPassword.passwordConfirmation')"
           name="Confirmation"
```
should we translate the name property as well?

carlobeltrame replied:
I removed the label and moved the translation to the name field now in efb4210998301a890f3056632f7010e82f9532a6

usu commented on frontend/src/views/auth/Register.vue:
```diff
@@ -2,98 +2,109 @@
-        dense
-        type="text"
-      />
+      <validation-observer>
```
If you want to use the validation-observer to prevent from submittal, the form needs to be be inside the validatio-observer. Same as done here for example: https://github.com/ecamp/ecamp3/blob/devel/frontend/src/views/CampCreate.vue#L4

As it is now, I can submit invalid forms which as soon as `formComplete === true`:

![image](https://user-images.githubusercontent.com/449555/194694032-ed9ab157-f6d3-43d7-afce-ef372ddeab24.png)

carlobeltrame replied:
Thanks, that did the trick (except for the checkbox)! Fixed in 867e7434fb9e979f998422eadb04ad950c592ed7

usu commented on frontend/src/components/form/base/EDatePicker.vue:
```diff
@@ -43,10 +44,12 @@ Displays a field as a date picker (can be used with v-model)
 <script>
 import BasePicker from './BasePicker.vue'
 import { HTML5_FMT } from '@/common/helpers/dateFormat.js'
+import { formComponentMixin } from '@/mixins/formComponentMixin.js'
```
This breaks the validation for the pickers:

**Why?** The validation is currently done on the textfield which is part of the picker. All properties unknown to the pickers are passed with `v-bind="$attrs"` from `EDatePicker` to `BasePicker` and from there to `ETextField`. Now that `veeId` and `veeRules` are known properties to `EDatePicker`, they are not passed automatically again.

**My proposal:** Revert the split of formComponentPropsMixin and formComponentMixin. That should work and still to the job for other components.

**How to test:** `CreateCampPeriods.vue` or `FormScheduleEntryItem.vue`

carlobeltrame replied:
The way you propose it, the red asterisk would stop working for the pickers. I think a better idea is to just pass these new props down to the BasePicker. The BasePicker will never notice any difference, and validation works great in my tests. Done in 5ed4afd7130f89aea539c96e3e41d4bd4208d33b

usu reviewed (changes requested):
Password strength checker looks nice 🤖 

But a few things we'd still need  to fix 😃

carlobeltrame committed:
Use a library which more closely models the OWASP recommendations for judging password quality

carlobeltrame committed:
Use name rather than label where possible

carlobeltrame committed:
Use vee-validate for preventing form submit where possible

I tried my best to get the checkbox validation to work, but without
deeply debugging vee-validate, I did not succees. So we still need the
formComplete computed, although it is far simpler now.

carlobeltrame committed:
Fix validation in picker components

carlobeltrame committed:
Avoid top-level await, it is not supported in jest as it seems

carlobeltrame committed:
Find a combination of props with which both vee-validate and vuetify are happy

Now the v-model value is correct, and validation-observer also works (as
long as we use vee-validate 3 syntax instead of older ones)

carlobeltrame committed:
Update snapshots now that the picker components support vee-validation

carlobeltrame committed:
Merge remote-tracking branch 'origin/devel' into login-email-validation

BacLuc commented on frontend/package.json:
```diff
@@ -44,6 +44,12 @@
+    "@zxcvbn-ts/language-it": "2.1.0",
     "assert": "2.0.0",
     "axios": "0.27.2",
     "check-password-strength": "2.0.7",
```
do we need that?

carlobeltrame committed:
Uninstall obsolete password strength checker package
